src/Features/Core/Portable/EditAndContinue/EditSession.cs

```diff
@@ -4,116 +4,145 @@
+        /// the state of the module when queried for the first time. Before we actually apply an edit to the module 
+        /// we need to query again instead of just reusing the diagnostic.
+        /// </summary>
+        private readonly Dictionary<Guid, ImmutableArray<LocationlessDiagnostic>> _moduleDiagnostics;
+        private readonly object _moduleDiagnosticsGuard = new object();
 
-        // EncEditSessionInfo is populated on a background thread and then read from the UI thread
-        private readonly object _encEditSessionInfoGuard = new object();
-        private EncEditSessionInfo _encEditSessionInfo = new EncEditSessionInfo();
+        /// <summary>
+        /// A <see cref="DocumentId"/> is added whenever <see cref="EditAndContinueDiagnosticAnalyzer"/> reports 
+        /// rude edits or module diagnostics. At the end of the session we ask the diagnostic analyzer to reanalyze 
+        /// the documents to clean up the diagnostics.
+        /// </summary>
+        private readonly HashSet<DocumentId> _documentsWithReportedDiagnostics;
+        private readonly object _documentsWithReportedDiagnosticsGuard = new object();
 
-        private readonly ImmutableDictionary<ActiveMethodId, ImmutableArray<NonRemappableRegion>> _nonRemappableRegions;
+        private bool _changesApplied;
 
-        internal EditSession(
-            Solution baseSolution,
-            DebuggingSession debuggingSession,
-            IActiveStatementProvider activeStatementProvider,
-            ImmutableDictionary<ProjectId, ProjectReadOnlyReason> projects,
-            ImmutableDictionary<ActiveMethodId, ImmutableArray<NonRemappableRegion>> nonRemappableRegions,
-            bool stoppedAtException)
+        internal EditSession(DebuggingSession debuggingSession, EditSessionTelemetry telemetry, bool stoppedAtException)
         {
-            Debug.Assert(baseSolution != null);
             Debug.Assert(debuggingSession != null);
-            Debug.Assert(activeStatementProvider != null);
-            Debug.Assert(nonRemappableRegions != null);
+            Debug.Assert(telemetry != null);
 
-            _baseSolution = baseSolution;
-            _debuggingSession = debuggingSession;
-            _activeStatementProvider = activeStatementProvider;
-            _stoppedAtException = stoppedAtException;
-            _projects = projects;
-            _cancellation = new CancellationTokenSource();
+            DebuggingSession = debuggingSession;
 
-            // TODO: small dict, pool?
-            _analyses = new Dictionary<DocumentId, Analysis>();
+            _analyses = new Dictionary<DocumentId, (Document Document, AsyncLazy<DocumentAnalysisResults> Results)>();
+            _documentsWithReportedDiagnostics = new HashSet<DocumentId>();
+            _moduleDiagnostics = new Dictionary<Guid, ImmutableArray<LocationlessDiagnostic>>();
```

Comment:
:bulb: Seems like these collections could be initialized with the declaration to match the guard object initializations.